converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Observation/ObservationParser.cs

```diff
@@ -0,0 +1,417 @@
+{
+    /// <summary>
+    /// Container for holding the parsed Observations from a DicomDataset
+    /// </summary>
+    public class ParsedObservation
+    {
+        public Collection<Observation> DoseSummaries { get; } = new();
+        public Collection<Observation> IrradiationEvents { get; } = new();
+    }
+
+    public static class ObservationParser
+    {
+        /// <summary>
+        /// Parses the DicomStructuredReports in a DicomDataset for the contained Observations.
+        /// </summary>
+        /// <remarks>
+        /// For observation profiles see: https://confluence.hl7.org/display/IMIN/Radiation+Dose+Summary+for+Diagnostic+Procedures+on+FHIR#RadiationDoseSummaryforDiagnosticProceduresonFHIR-DoseSummaryObservationProfile
+        /// </remarks>
+        /// <param name="dataset">The dataset to parse</param>
+        /// <param name="patientRef">Reference to an existing patient to link to</param>
+        /// <param name="imagingStudyRef">Reference to an existing ImagingStudy to link to</param>
+        /// <returns>An object containing the Observations found in the dataset</returns>
+        public static ParsedObservation CreateObservations(
+            DicomDataset dataset,
+            ResourceReference patientRef,
+            ResourceReference imagingStudyRef)
+        {
+            var parsedObservations = new ParsedObservation();
+
+            // Recursive loop to add all found Dose Summaries and Irradiation Events to the passed observation set
+            static void ObservationParseLoop(
+                DicomDataset dataset,
+                ResourceReference patientRef,
+                ResourceReference imagingStudyRef,
+                ParsedObservation observations
+            )
+            {
+                var report = new DicomStructuredReport(dataset);
+                // see if the current report code matches any of the observation container codes; if so, create the appropriate observation
+                try
+                {
+                    (string Value, string Scheme) lookupTuple = (report.Code.Value, report.Code.Scheme);
+                    if (DoseSummaryReportCodes.Contains(lookupTuple))
+                    {
+                        Observation doseSummary = CreateDoseSummary(
+                            dataset,
+                            imagingStudyRef,
+                            patientRef);
+                        observations.DoseSummaries.Add(doseSummary);
+                    }
+
+                    if (IrradiationEventCodes.Contains(lookupTuple))
+                    {
+                        Observation irradiationEvent = CreateIrradiationEvent(dataset, patientRef);
+                        observations.IrradiationEvents.Add(irradiationEvent);
+                    }
+                }
+                catch (MissingMemberException)
+                {
+                    // Occurs when a required attribute is unable to be extracted from a dataset.
```

Comment:
should we log anything here?

Reply:
Possibly. We define some dicom attributes as being required for their perspective FHIR observation profiles (e.g. Dose summary FHIR profile requires a `StudyInstanceUID` and `AccessionNumber`). But I do not know if those fields are actually required in the dicom spec itself. If they are, then I'd say it makes sense to log something. But if its expected that those fields are optional; it might be just be unnecessary noise. Thoughts?